server.js

```diff
@@ -1,30 +1,58 @@
-// Log memory usage before doing GC. Seems to blow up the file limit sooner than without.
-var gcLog = function () {
-  console.log(new Date().toISOString() + JSON.stringify(process.memoryUsage()));
-  gc();
-};
+var fs = require('fs');
+var ws = require('ws');
+var http = require('http');
+var randomBytes = require('crypto').randomBytes;
 
-// Tell original client how many clients are connected
-var reporter = function () {
-  if (this.clients[0]) this.clients[0].send(this.clients.length.toString());
-};
+var html = fs.readFileSync('./index.html');
+var browsers = [];
 
-// Garbage collect every 30 seconds
-setInterval(gc, 30 * 1000);
+var www = http.createServer(function (req, res) {
+  res.writeHead(200, { 'Content-Type': 'text/html' });
+  res.end(html);
+}).listen(process.env.PORT || 5000);
 
-// Create the WS server
-new (require('ws').Server)({
+var wss = new ws.Server({
+  server: www
+}).on('connection', function (socket) {
+  socket.on('message', onmessage);
+  socket.on('close', logSocketCount);
+  logSocketCount();
+});
 
-  // Attach it to an HTTP server that serves a tiny web client
-  server: require('http').createServer(function (req, res) {
+// Log memory usage before doing GC. Seems to blow up the file limit sooner than without.
+setInterval(logMemoryUsage, 30 * 1000);
 
-    res.writeHead(200, { 'Content-Type': 'text/html' });
-    res.end('<script>new WebSocket("ws://"+location.host).onmessage=function(msg){document.body.innerText=msg.data;};</script>');
+function logMemoryUsage () {
+  console.log(new Date().toISOString() + JSON.stringify(process.memoryUsage()));
+  gc();
+}
 
-  }).listen(process.env.PORT || 5000)
+function getSocketCount () {
+  return wss.clients.length.toString();
+}
 
-// Call the reporter on connection and close events
-}).on('connection', function (connection) {
-  reporter.call(this);
-  connection.on('close', reporter.bind(this));
-});
+function logSocketCount () {
+  var n = getSocketCount();
+  browsers.forEach(function (b) {
+    b.readyState === 1 && b.send(n);
+  });
+  console.log(n);
+}
+
+function onmessage (message) {
+  if (message === 'browser') {
+    this.removeListener('message', onmessage);
+    browsers.push(this);
+    this.on('close', function () {
+      browsers.splice(browsers.indexOf(this), 1);
+    });
+    this.send(getSocketCount());
+  } else {
+    var self = this;
+    setTimeout(function () {
+      if (self.readyState === 1) {
+        self.send(randomBytes(message.length));
+      }
+    }, Math.random() * 1000);
```

Comment:
If you pass the --messages <n> flag to thor, it will send n messages of size --buffer <s> on each connection. A response is expected before each thor connection will continue on to sending it's next message - once it has received and sent n messages, the connection is torn down. I figured responding with a message of the same size as the one thor sent us seemed appropriate, and the call to randomBytes has us doing at least some "work" to generate the message.